### PR TITLE
feat(ops): optional OpState

### PR DESCRIPTION
**Pull request by AaronO**
Following https://github.com/denoland/deno/pull/13861, part of https://github.com/denoland/deno/issues/13943

### CODE REVIEW

File: ops/lib.rs [hunk 1]
```diff
@@ -136,15 +145,23 @@ fn codegen_v8_async(core: &TokenStream2, f: &syn::ItemFn) -> TokenStream2 {
...
-  let (arg_decls, args_tail) = codegen_args(core, f, 1, 1);
+  let arg0 = f.sig.inputs.first();
+  let uses_opstate = arg0.map(is_mut_ref_opstate).unwrap_or_default();
+  let args_head = if uses_opstate {
+    quote! { op_state, }
+  } else {
+    quote! {}
+  };
+  let rust_i0 = if uses_opstate { 1 } else { 0 };
+  let (arg_decls, args_tail) = codegen_args(core, f, rust_i0, 1);
```
Comment (littledivy) [inline ops/lib.rs:164]:
```suggestion
  let (arg_decls, args_tail) = codegen_args(core, f, usize::from(uses_opstate), 1);
```

File: ops/lib.rs [hunk 2]
```diff
@@ -91,7 +92,15 @@ pub fn op(_attr: TokenStream, item: TokenStream) -> TokenStream {
...
-  let (arg_decls, args_tail) = codegen_args(core, f, 1, 2);
+  let arg0 = f.sig.inputs.first();
+  let uses_opstate = arg0.map(is_rc_refcell_opstate).unwrap_or_default();
+  let args_head = if uses_opstate {
+    quote! { state, }
+  } else {
+    quote! {}
+  };
+  let rust_i0 = if uses_opstate { 1 } else { 0 };
+  let (arg_decls, args_tail) = codegen_args(core, f, rust_i0, 2);
```
Comment (littledivy) [inline ops/lib.rs:103]:
```suggestion
  let (arg_decls, args_tail) = codegen_args(core, f, usize::from(uses_opstate), 2);
```

File: ops/lib.rs (hunk 1 above)
Comment (AaronO) [reply ops/lib.rs:164]: I think coercing bools to arg offsets could hinder maintainability more than the verbosity of the if expr

File: ops/lib.rs (hunk 2 above)
Comment (AaronO) [reply ops/lib.rs:103]: I think coercing bools to arg offsets could hinder maintainability more than the verbosity of the if expr

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by AaronO** [comment]
> LGTM, defining ops feels a bit magical, but it's a good net save in source lines 👍

It's always been magical ... ✨

**Pull request merged**